packages/react-server/src/ReactFlightServer.js

```diff
@@ -1923,20 +1916,23 @@ function visitAsyncNode(
       const awaited = node.awaited;
       let match = null;
       if (awaited !== null) {
-        const ioNode = visitAsyncNode(request, task, awaited, cutOff, visited);
+        const ioNode = visitAsyncNode(request, task, awaited, visited);
         if (ioNode !== null) {
           // This Promise was blocked on I/O. That's a signal that this Promise is interesting to log.
           // We don't log it yet though. We return it to be logged by the point where it's awaited.
           // The ioNode might be another PromiseNode in the case where none of the AwaitNode had
           // unfiltered stacks.
-          if (
+          if (ioNode.tag === PROMISE_NODE) {
+            // If the ioNode was a Promise, then that means we found one in user space since otherwise
+            // we would've returned an IO node. We assume this has the best stack.
+            match = ioNode;
```

Comment:
This fixes the issue where the ThirdPartyComponent's `delay` call showed up as `ThirdPartyComponent` as the I/O's name instead of `delay`.

That was because if there's no explicit `await` then we picked the outer most Promise as the stack. That made no sense since that could just be the top component. We want the inner one and we hope that it has an "async stack" which includes higher up abstractions as well.